fw/http_parser.c

```diff
@@ -6044,7 +6058,7 @@ __h2_req_parse_authority(TfwHttpReq *req, unsigned char *data, size_t len,
 	__FSM_STATE(Req_I_A_v6) {
 		/* See Req_UriAuthorityIPv6 processing. */
 		if (likely(isxdigit(c) || c == ':'))
-			__FSM_H2_I_MOVE_fixup(Req_I_A_v6, 1, TFW_STR_VALUE);
+			__FSM_H2_I_MOVE_NEQ_n_flag(Req_I_A_v6, 1, TFW_STR_VALUE);
```

Comment:
Which exactly tests are broken without the fixup-removing change?

At the moment I'm confused with `__msg_hdr_chunk_fixup()` at the end of parsing and it seems for me that the code works for contiguous input data only (note that the current chunking tests aren't ideal since they just splits a contiguous data and chunks are actually not bordered).

Reply:
So what is wrong with the scenario `authority: [::1]:5001` from #1566? You do the opposite change, so we need a very precise understanding on what was wrong with that change, i.e. were the conclusions about the test case processing were wrong.

Reply:
OK, looks good. My understanding of https://github.com/tempesta-tech/tempesta/pull/1566/commits/0d98a42425f2b074a21bd50deea25926aa83de34 is that to prevent `[[::1]:5001` parsed from 2 chunks `[:`, `:1]:5001` we shouldn't fixup `[` in `Req_I_A_Start`.

In general, explicit fixups were introduced to explicitly separate contiguous strings, e.g. `Host: foo.com`, to separate parts for easier operation, in this case there header name and the header value. In normal case we fixup a chunk only when there is not enough data (returning TFW_POSTPONE). We should not create extra chunks because this hurt performance. So the statement that we either use fixup or non-fixup functions is wrong.

Reply:
> We should not create extra chunks because this hurt performance. So the statement that we either use fixup or non-fixup functions is wrong.

Yes, I totally agreed with this a couple of posts ago:

> This version of code (before my changes) can be fixed with such changes in the last step of parsing:
> ```
> 		if (likely(c == ']')) {
> 			__msg_hdr_chunk_fixup(p, 1);
> 			__msg_chunk_flags(TFW_STR_HDR_VALUE | TFW_STR_VALUE);
> ```
> But this version of the code with _fixup-macros has one disadvantage: we always fixup data by one chunk at a time, no matter what the actual size of the chunks is.